src/projections/adams.cpp

```diff
@@ -158,24 +160,122 @@ static PJ_XY forward(PJ_LP lp, PJ *P) {
         break;
     }
 
-    double m = aasin(P->ctx, sqrt((1. + cos(a + b))));
+    double m = aasin(P->ctx, sqrt((1. + std::min(0.0, cos(a + b)))));
     if (sm) m = -m;
 
-    double n = aasin(P->ctx, sqrt(fabs(1. - cos(a - b))));
+    double n = aasin(P->ctx, sqrt(fabs(1. - std::max(0.0, cos(a - b)))));
     if (sn) n = -n;
 
     xy.x = ell_int_5(m);
     xy.y = ell_int_5(n);
 
     if (Q->mode == ADAMS_HEMI || Q->mode == ADAMS_WS2) { /* rotate by 45deg. */
-        double temp = xy.x;
+        const double temp = xy.x;
         xy.x = RSQRT2 * (xy.x - xy.y);
         xy.y = RSQRT2 * (temp + xy.y);
     }
 
     return xy;
 }
 
+static PJ_LP inverse_with_newton_raphson(PJ_XY xy,
```

Comment:
Just thinking out loud here. This would be a nice thing to do for all projections without an inverse. We could make this a fallback in `pj_inv()` in the case no proper inverse function is registered with the projection. I think all that it would take is a mechanism to provide the initial guess. There may be ways to automate that as well, for instance by caching previous inverse coordinates and other hacks.

Reply:
yes, this has the potential to be generic but a number of error thresholds in it have been tuned to make it work with adams_ws2 (for example it somehow assumes that (0,0) maps to (0,0), etc etc) . So for now, it is well here and we can potentially try to reuse it in a more general case if we need

Reply:
I have a bit of extra time on my hands next week. I think I will give it a first go and try to uncover some of the pitfalls. If we can make it work we can increase the amount of projections with an inverse quite significantly!